From c32d308cf5ec9b6ae7d6aed7a1eee50d16ee8add Mon Sep 17 00:00:00 2001
From: Matteo Bertozzi <matteo.bertozzi@cloudera.com>
Date: Wed, 30 Jan 2013 16:42:24 -0800
Subject: [PATCH 195/202] HBASE-7684 NullPointerException in SecureClient when Call is cleaned up due to RPC timeout

Reason: Bug
Author: cuijianwei
Ref: CDH-10220
---
 .../org/apache/hadoop/hbase/ipc/SecureClient.java  |   12 +++++++++---
 1 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/security/src/main/java/org/apache/hadoop/hbase/ipc/SecureClient.java b/security/src/main/java/org/apache/hadoop/hbase/ipc/SecureClient.java
index 4c7bcbb..ce8305f 100644
--- a/security/src/main/java/org/apache/hadoop/hbase/ipc/SecureClient.java
+++ b/security/src/main/java/org/apache/hadoop/hbase/ipc/SecureClient.java
@@ -375,10 +375,16 @@ public class SecureClient extends HBaseClient {
           if (LOG.isDebugEnabled()) {
             LOG.debug("call #"+id+", response is:\n"+value.toString());
           }
-          call.setValue(value);
+          // it's possible that this call may have been cleaned up due to a RPC
+          // timeout, so check if it still exists before setting the value.
+          if (call != null) {
+            call.setValue(value);
+          }
         } else if (state == Status.ERROR.state) {
-          call.setException(new RemoteException(WritableUtils.readString(in),
-                                                WritableUtils.readString(in)));
+          if (call != null) {
+            call.setException(new RemoteException(WritableUtils.readString(in), WritableUtils
+                .readString(in)));
+          }
         } else if (state == Status.FATAL.state) {
           // Close the connection
           markClosed(new RemoteException(WritableUtils.readString(in),
-- 
1.7.0.4

